### PR TITLE
Add custom type hinting

### LINKED ISSUE #370: Add missing type hints
A few functions definitions are missing type hints.

* `merkle_root`
* `to_penalize`
* `deposit`
* `get_ancestor`
* `lmd_ghost`
* `get_vote_count`
* `hash`
* `xor`

### PR DESCRIPTION
**Pull request by hwwhww**
Fix #370 except for the fork choice rule section:

The `lmd_ghost` has a highly abstracted `store` and `start` objects. For correctness, we *can* define the `TStore = TypeVar("Store")` and `TStart = TypeVar("Start")` to deal with it. Would that be "too Pythonic" to our readers?

### CODE REVIEW

**Comment by JustinDrake** [comment]
For `lmd_ghost` we don't need `start`, only `start.state` which is of type `BeaconState`.

**Comment by djrtwo** [comment]
We need the associated `block` related to the `start_state`. Looks like the algorithm is implicitly stating that `start` is the block at that start state (genesis state or last finalized block), and `start.state` is the associated state.

Might be clearer to pull these how at separate vars `start_block`, `start_state` (or something similar), add types for both and add a docstring.

**Comment by djrtwo** [comment]
@hwwhww Are you going to take care of making the two args (block and state) explicit with typing? I'm happy to take care of it if you want

**Comment by hwwhww** [comment]
@JustinDrake @djrtwo I updated it:
`def lmd_ghost(store: Store, start_state: BeaconState, start_block: BeaconBlock) -> BeaconBlock`.

#### Changelog:
* PR feedback: update `lmd_ghost` params and add description of `justified_head_state`.
* Add an abstract type `Store` for making type hinting completed.
* Add definition of `get_children` and add `store: Store` into the parameters.

Feel free to add commits to fix bugs or wording. 🙏

File: specs/core/0_beacon-chain.md [hunk 1]
```diff
@@ -611,6 +614,20 @@ Code snippets appearing in `this style` are to be interpreted as Python code. Be
...
+## Custom Types
+
+We define the following Python custom types for type hinting and readability:
+
+| Name | Type | Description |
+| - | - | - |
+| `SlotNumber` | unsigned 64-bit integer | the number of a slot |
+| `ShardNumber` | unsigned 64-bit integer | the number of a shard |
+| `ValidatorIndex` | unsigned 24-bit integer | the index number of validator in the registry |
+| `Gwei` | unsigned 64-bit integer | the amount in Gwei |
```
Comment (djrtwo) [inline specs/core/0_beacon-chain.md]: Here are a few tiny suggestions that sound more natural imo
```suggestion
| `Gwei` | unsigned 64-bit integer | an amount in Gwei |
```

File: specs/core/0_beacon-chain.md [hunk 2]
```diff
@@ -611,6 +614,20 @@ Code snippets appearing in `this style` are to be interpreted as Python code. Be
...
 
+## Custom Types
+
+We define the following Python custom types for type hinting and readability:
+
+| Name | Type | Description |
+| - | - | - |
+| `SlotNumber` | unsigned 64-bit integer | the number of a slot |
+| `ShardNumber` | unsigned 64-bit integer | the number of a shard |
+| `ValidatorIndex` | unsigned 24-bit integer | the index number of validator in the registry |
```
Comment (djrtwo) [inline specs/core/0_beacon-chain.md]:
```suggestion
| `ValidatorIndex` | unsigned 24-bit integer | the index number of a validator in the registry |
```

File: specs/core/0_beacon-chain.md [hunk 3]
```diff
@@ -611,6 +614,20 @@ Code snippets appearing in `this style` are to be interpreted as Python code. Be
...
+| `Gwei` | unsigned 64-bit integer | the amount in Gwei |
+| `Bytes32` | 32-byte data | the binary data in 32-byte length |
```
Comment (djrtwo) [inline specs/core/0_beacon-chain.md]:
```suggestion
| `Bytes32` | 32-byte data | binary data with 32-byte length |
```

File: specs/core/0_beacon-chain.md [hunk 4]
```diff
@@ -611,6 +614,20 @@ Code snippets appearing in `this style` are to be interpreted as Python code. Be
...
+| `BLSPubkey` | 48-byte data | the public key in BLS signature scheme |
```
Comment (djrtwo) [inline specs/core/0_beacon-chain.md]:
```suggestion
| `BLSPubkey` | 48-byte data | a public key in BLS signature scheme |
```

File: specs/core/0_beacon-chain.md [hunk 5]
```diff
@@ -611,6 +614,20 @@ Code snippets appearing in `this style` are to be interpreted as Python code. Be
...
+| `BLSSignature` | 96-byte data | the signature in BLS signature scheme |
```
Comment (djrtwo) [inline specs/core/0_beacon-chain.md]:
```suggestion
| `BLSSignature` | 96-byte data | a signature in BLS signature scheme |
```

File: specs/core/0_beacon-chain.md [hunk 6]
```diff
@@ -729,28 +746,38 @@ Beacon block production is significantly different because of the proof of stake
 
 The beacon chain fork choice rule is a hybrid that combines justification and finality with Latest Message Driven (LMD) Greediest Heaviest Observed SubTree (GHOST). At any point in time a [validator](#dfn-validator) `v` subjectively calculates the beacon chain head as follows.
 
-* Let `store` be the set of attestations and blocks that the [validator](#dfn-validator) `v` has observed and verified (in particular, block ancestors must be recursively verified). Attestations not part of any chain are still included in `store`.
+* Abstractly define `Store` as the type of storage object for the chain data and `store` be the set of attestations and blocks that the [validator](#dfn-validator) `v` has observed and verified (in particular, block ancestors must be recursively verified). Attestations not part of any chain are still included in `store`.
```
Comment (djrtwo) [inline specs/core/0_beacon-chain.md]:
```suggestion
* Abstractly define `Store` as the type of storage object for the chain data and `store` be the set of attestations and blocks that the [validator](#dfn-validator) `v` has observed and verified (in particular, block ancestors must be recursively verified). Attestations not yet included in any chain are still included in `store`.
```

File: specs/core/0_beacon-chain.md [hunk 7]
```diff
@@ -729,28 +746,38 @@ Beacon block production is significantly different because of the proof of stake
...
-* The head is `lmd_ghost(store, justified_head)` where the function `lmd_ghost` is defined below. Note that the implementation below is suboptimal; there are implementations that compute the head in time logarithmic in slot count.
+* Let `get_ancestor(store: Store, block: BeaconBlock, slot: SlotNumber) -> BeaconBlock` be the ancestor of `block` with slot number `slot`. The `get_ancestor` function can be defined recursively as `def get_ancestor(store: Store, block: BeaconBlock, slot: SlotNumber) -> BeaconBlock: return block if block.slot == slot else get_ancestor(store, store.get_parent(block), slot)`.
+* Let `get_latest_attestation(store: Store, validator: Validator) -> Attestation` be the attestation with the highest slot number in `store` from `validator`. If several such attestations exist, use the one the [validator](#dfn-validator) `v` observed first.
+* Let `get_latest_attestation_target(store: Store, validator: Validator) -> BeaconBlock` be the target block in the attestation `get_latest_attestation(store, validator)`.
+* Let `get_children(store: Store, block: BeaconBlock) -> List[BeaconBlock]` returns the children blocks of the given `block`.
```
Comment (djrtwo) [inline specs/core/0_beacon-chain.md]:
```suggestion
* Let `get_children(store: Store, block: BeaconBlock) -> List[BeaconBlock]` returns the child blocks of the given `block`.
```

File: specs/core/0_beacon-chain.md [hunk 8]
```diff
@@ -729,28 +746,38 @@ Beacon block production is significantly different because of the proof of stake
...
+* Let `justified_head_state` be the `BeaconState` object after processing `justified_head`.
```
Comment (djrtwo) [inline specs/core/0_beacon-chain.md]:
```suggestion
* Let `justified_head_state` be the resulting `BeaconState` object from processing the chain up to the `justified_head`.
```

**Pull request merged**